File: systemid.cpp
//
// Includes
//

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <random>
#include <Eigen/Dense>

//
// Namespaces
//

using namespace std;
using namespace Eigen;

//
// Constants
//

// A weights (3x3 matrix)
const MatrixXd A
{
  { -0.1199, 2.0558, 1.1026 },
  { -1.8226, -0.7465, -15.0483 },
  { 0.7375,  8.2928,  -16.6951 }
};

// B weights (3x1 vector)
const RowVectorXd B {{ 0.0009, 0.1359, 0.3223 }};

// C weights (1x3 vector)
const VectorXd C {{ -32.1738, -0.7282, 0.1308 }};

// D weight (scalar)
const double D = 8.7666e-04;

// K weights (3x1 vector)
const RowVectorXd K {{ -1.3722, -4.1560, 12.5563 }};

// x0 initial state (3x1 vector)
const RowVectorXd x0 {{ 0.0, 0.0, 0.0 }};

// Disturbance
const double DISTURBANCE_MEAN = 0.0;
const double DISTURBANCE_VAR = 841.9616;

//
// Functions
//

// Cx + Du + e
double systemModel(
  const RowVector3d& x, double u, double e)
{
  return
    // Add contribution of state
    (C * x).sum() +
    // Add contribution of input
    D * u +
    // Add disturbance
    e;
}

// Ax + Bu + Ke
RowVectorXd systemState(
  const RowVectorXd& x, double u, double e)
{
  return
    // Add contribution of state
    A * x +
    // Add contribution of input
    B * u +
    // Add contribution of disturbance
    K * e;
}

bool openOutput(const string& path, ofstream& file)
{
  file.open(path);
  return file.is_open();
}

bool openInput(const string& path, ifstream& file)
{
  file.open(path);
  return file.is_open();
}

int main(int argc, char** argv)
{
  // Open input
  ifstream input;

  if (!openInput("input.csv", input))
  {
    cerr << "Failed to open input file" << endl;
    return 1;
  }

  // Open output
  ofstream output;

  if (!openOutput("output.csv", output))
  {
    cerr << "Failed to open output file" << endl;
    return 1;
  }

  // Disturbance
  random_device rd;
  mt19937 gen(rd());
  normal_distribution<double> dist(
    DISTURBANCE_MEAN, DISTURBANCE_VAR);

  // Kalman filter
  RowVectorXd x = x0;
  string line;
  double lastTime = 0.0;

  // Skip headers
  getline(input, line);

  while(!input.eof())
  {
    // Read time, input, output
    double time, u, y;
    getline(input, line);
    sscanf(line.c_str(), "%lf, %lf, %lf", &time, &u, &y);

    // Calculate time step
    double timeStep = lastTime - time;
    lastTime = time;

    // Predict
    double estimate = systemModel(x, u, 0.0);

    // Update state
    x = systemState(x, u, 0.0);

    output << estimate << endl;
  }

  return 0;
}

File: systemid.m
% A weights (3x3 matrix)
A = [ ...
  0.0356,   -3.4131,  -14.9525;
  -1.0591,   85.8128,  334.3098;
  1.5729,  -95.1123, -175.5517;
];

% B weights (3x1 vector)
B = [ ...
  -0.0019;
  0.0403;
  -0.0225;
];

% C weights (1x3 vector)
C = [ -5316.9, 24.87, 105.92 ];

% D weight (scalar)
D = 0.05;

% K weights (3x1 vector)
K = [ ...
  -0.0025;
  -0.0582;
  0.0984;
];

% Initial state
x0 = [ ...
  -0.0458;
  0.0099;
  -0.0139;
];

% Read input file
csv = readmatrix('input.csv');

% Select column vectors
time = csv(:,1);
measurement = csv(:,2);
input = csv(:,3);

% Simulate
x = x0;
output = zeros(length(time), 1);

for i = 1:length(time)
  if i > 1
    timeStep = time(i) - time(i-1);
  else
    timeStep = 0;
  end

  u = input(i);
  y = systemModel(x, u, 0, C, D);
  x = systemState(x, u, 0, A, B, K) * timeStep;

  output(i) = y;
end

function prediction = systemModel(x, u, e, C, D)
  prediction = ...
    sum(C * x) + ...  % Add contribution of state
    D * u + ...       % Add contribution of input
    e;                % Add contribution of disturbance
end

function x1 = systemState(x0, u, e, A, B, K)
  x1 = ...
    A * x0 + ...      % Add contribution of state
    B * u + ...       % Add contribution of input
    e * K;            % Add contribution of disturbance
end
